bcipy/language/model/gpt2.py

```diff
@@ -0,0 +1,188 @@
+from typing import List, Tuple, Dict
+from bcipy.helpers.task import SPACE_CHAR, BACKSPACE_CHAR
+from bcipy.language.main import LanguageModel, ResponseType
+
+import torch
+import torch.nn.functional as F
+from transformers import GPT2LMHeadModel, GPT2Tokenizer
+from collections import Counter
+
+
+class GPT2LanguageModel(LanguageModel):
+
+    def __init__(self, response_type, symbol_set, lm_path):
+        """
+        Initilize instance variables and load the language model with given path
+        Args:
+            response_type - SYMBOL only
+            symbol_set - list of symbol strings
+            lm_path - path to language model files
+        """
+        self.response_type = response_type
+        self.symbol_set = symbol_set
+        self.__validate_response_type()
+        self.model = None
+        self.tokenizer = None
+        self.is_start_of_word = True
+        self.vocab_size = 0
+        self.idx_to_word = None
+        self.curr_word_predicted_prob = None
+        self.lm_path = lm_path
+        self.load()
+
+    def __validate_response_type(self):
+        """
+        Make sure the response_type is valid for this language model
+        """
+        assert self.response_type == ResponseType.SYMBOL, "response type must be SYMBOL!"
+
+    def __get_char_predictions(self, word_prefix: str) -> List[tuple]:
+        """
+        Given a word prefix typed by user, predict the probability distribution
+        of next character
+        Args:
+            word_prefix - string
+        Response:
+            A list of tuples - each tuple consists a pair of symbol and probability
+        """
+
+        prefix_len = len(word_prefix)
+        char_to_prob = Counter()
+
+        # iterate over the entire vocabulary
+        for idx in range(self.vocab_size):
+            # change word and word_prefix to upper case
+            word = self.idx_to_word[idx].upper()
+            word_prefix = word_prefix.upper()
+
+            # if a word in the vocabulary matches the word prefix
+            if word.startswith(word_prefix):
+                # obtain the character following the prefix in this word
+                if len(word) == prefix_len:
+                    next_char = SPACE_CHAR
+                else:
+                    next_char = word[prefix_len]
+                # make sure the next character is in the symbol set
+                if next_char in self.symbol_set and next_char != BACKSPACE_CHAR:
+                    char_to_prob[next_char] += self.curr_word_predicted_prob[idx]
+
+        # normalize char_to_prob
+        sum_char_prob = sum(char_to_prob.values())
+        for char in char_to_prob:
+            char_to_prob[char] /= sum_char_prob
+
+        # build a list of tuples (char, prob)
+        char_prob_tuples = [(k, v) for k, v in sorted(char_to_prob.items(), key=lambda item: item[1], reverse=True)]
```

Comment:
You don't need the list comprehension here:

    list(sorted(char_to_prob.items(), key=lambda item: item[1], reverse=True))

Reply:
Yes, I will change this.